src/plugins/telemetry/server/telemetry_collection/ingest_solutions/ingest_solutions.ts

```diff
@@ -0,0 +1,177 @@
+/*
+ * Licensed to Elasticsearch B.V. under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import { APICaller } from 'kibana/server';
+import { INGEST_SOLUTIONS } from './constants';
+
+type TechnologyDataProviders = typeof INGEST_SOLUTIONS[number]['name'];
+
+export interface IngestSolutionsPayload {
+  data_providers: {
+    [key in TechnologyDataProviders]?: {
+      index_count: number;
+      ecs_index_count?: number;
+      doc_count?: number;
+      size_in_bytes?: number;
+    };
+  };
+}
+
+export interface IngestSolutionsIndex {
+  name: string;
+  isECS?: boolean; // Optional because it can't be obtained via Monitoring.
+
+  // The fields below are optional because we might not be able to obtain them if the user does not
+  // have access to the index.
+  docCount?: number;
+  sizeInBytes?: number;
+}
+
+export function buildIngestSolutionsPayload(
```

Comment:
This made my head hurt. Is there any chance of simplifying the nested use of reduce here? I ask because it will be easier to jump back into the code 6 months from now for a dev who didn't write the code. We could then also extend unit test coverage, should we need to.

Reply:
I've taken a more _functional_ approach. Moving some of the logic inside the `.reduce` to separate functions. I hope it helps with readability.
But if you think the `.reduce` approach makes things harder to read, I can use a `for...in` loop instead:

```typescript
export function buildIngestSolutionsPayload(
  indices: IngestSolutionsIndex[]
): IngestSolutionsPayload {
  const startingDotPatternsUntilTheFirstAsterisk = INGEST_SOLUTIONS.map(({ pattern }) =>
    pattern.replace(/^\.(.+)\*.*$/g, '.$1')
  ).filter(Boolean);

  // Filter out the system indices unless they are required by the patterns
  const indexCandidates = indices.filter(
    ({ name }) =>
      !(
        name.startsWith('.') &&
        !startingDotPatternsUntilTheFirstAsterisk.find(pattern => name.startsWith(pattern))
      )
  );

  const ingestSolutionsPayload = {} as IngestSolutionsPayload;
  for (const indexCandidate of indexCandidates) {
    const matchingTechnology = findMatchingTechnology(indexCandidate.name);

    if (matchingTechnology) {
      if (!ingestSolutionsPayload.data_providers) {
        ingestSolutionsPayload.data_providers = {};
      }

      const { name } = matchingTechnology;
      ingestSolutionsPayload.data_providers[name] = increaseCounters(
        ingestSolutionsPayload.data_providers[name],
        indexCandidate
      );
    }
  }
  return ingestSolutionsPayload;
}
```

Personally, I'm not a big fan of `for` loops that mutate external objects (`ingestSolutionsPayload`) and I always try to avoid them when possible. I think it opens the possibility (and may even encourage) to create other loops to keep mutating them, making it harder to trace and to anticipate the result. 
On top of that, I'd say it doesn't really reduce the level of nesting. On the contrary, since you can't use `return` statements, it may even increase them to wrap them inside `if` statements (as seen in this example). We have `continue` though... 😅 

Nevertheless, if you think the `for...of` loop improves readability. I can def change it :)

Reply:
No need to change from the `reduce`. The functional approach is great!